internal/grpcutil/metadata.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcutil
+
+import (
+	"context"
+	"strings"
+
+	v1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// ConvertMetadataToProtoHeader converts gRPC metadata into proto Headers.
+func ConvertMetadataToProtoHeader(
+	src metadata.MD,
+) []*v1alpha1.Header {
+	headerInfo := make([]*v1alpha1.Header, 0, len(src))
+	for key, value := range src {
+		hdr := &v1alpha1.Header{
+			Name:  key,
+			Value: value,
+		}
+		headerInfo = append(headerInfo, hdr)
+	}
+	return headerInfo
+}
+
+// ConvertProtoHeaderToMetadata converts a slice of proto Headers into gRPC metadata.
+func ConvertProtoHeaderToMetadata(
+	src []*v1alpha1.Header,
+) metadata.MD {
+	md := make(metadata.MD, len(src))
+	for _, hdr := range src {
+		key := strings.ToLower(hdr.Name)
+		md[key] = hdr.Value
+	}
+	return md
+}
+
+// AddHeaderMetadata adds the given proto Headers into the context.
```

Comment:
FWIW, this does not add the given headers to the context. This is actually adding the response headers to the current RPC. Maybe that's a nit, but in my head, the existing wording is confusing since they're not actually stored in the context. (What's actually happening is that the current stream is stored in the context, and `grpc.SetHeader` and `grpc.SetTrailer` pull the stream out of context and store the metadata there, so it can be written to the client later.)

Reply:
Gotcha. Updated the comments.